library/std/src/sys/uefi/common.rs

```diff
@@ -0,0 +1,515 @@
+pub(crate) struct ProtocolWrapper<P>
+where
+    P: Protocol,
+{
+    handle: NonNull<crate::ffi::c_void>,
+    protocol: Box<P>,
+}
+
+impl<P> ProtocolWrapper<P>
+where
+    P: Protocol,
+{
+    #[inline]
+    pub(crate) fn new(handle: NonNull<crate::ffi::c_void>, protocol: Box<P>) -> Self {
+        Self { handle, protocol }
+    }
+
+    pub(crate) fn install_protocol(protocol: P) -> io::Result<ProtocolWrapper<P>> {
+        let mut handle: r_efi::efi::Handle = crate::ptr::null_mut();
+        let mut protocol = Box::new(protocol);
+        install_protocol::<P>(&mut handle, P::PROTOCOL_GUID, &mut protocol)?;
+        let handle = NonNull::new(handle)
+            .ok_or(io::const_io_error!(io::ErrorKind::Uncategorized, "found null handle"))?;
+        Ok(Self::new(handle, protocol))
+    }
+
+    pub(crate) fn install_protocol_in(
+        protocol: P,
+        mut handle: r_efi::efi::Handle,
+    ) -> io::Result<ProtocolWrapper<P>> {
+        let mut protocol = Box::new(protocol);
+        install_protocol::<P>(&mut handle, P::PROTOCOL_GUID, &mut protocol)?;
+        let handle = NonNull::new(handle)
+            .ok_or(io::const_io_error!(io::ErrorKind::Uncategorized, "found null handle"))?;
+        Ok(Self::new(handle, protocol))
+    }
+
+    #[inline]
+    pub(crate) fn handle(&self) -> NonNull<crate::ffi::c_void> {
+        self.handle
+    }
+}
+
+impl<P> Drop for ProtocolWrapper<P>
+where
+    P: Protocol,
+{
+    #[inline]
+    fn drop(&mut self) {
+        let _ = uninstall_protocol::<P>(self.handle.as_ptr(), P::PROTOCOL_GUID, &mut self.protocol);
+    }
+}
+
+/// Get the BootServices Pointer.
+pub(crate) fn get_boot_services() -> Option<NonNull<r_efi::efi::BootServices>> {
+    let system_table: NonNull<r_efi::efi::SystemTable> = uefi::env::system_table()?.cast();
+    let boot_services = unsafe { (*system_table.as_ptr()).boot_services };
+    NonNull::new(boot_services)
+}
+
+/// Get the RuntimeServices Pointer.
+pub(crate) fn get_runtime_services() -> Option<NonNull<r_efi::efi::RuntimeServices>> {
+    let system_table: NonNull<r_efi::efi::SystemTable> = uefi::env::system_table()?.cast();
+    let runtime_services = unsafe { (*system_table.as_ptr()).runtime_services };
+    NonNull::new(runtime_services)
+}
+
+// Create UCS-2 Vector from OsStr
+pub(crate) fn to_ffi_string(s: &OsStr) -> Vec<u16> {
+    let mut v: Vec<u16> = s
+        .encode_wide()
+        .map(|x| match x {
+            Ok(c) => c,
+            Err(_) => u16::from(crate::sys_common::ucs2::Ucs2Char::REPLACEMENT_CHARACTER),
+        })
+        .collect();
+    v.push(0);
+    v.shrink_to_fit();
+    v
+}
+
+// Create OsString from UEFI UCS-2 String
+pub(crate) fn from_ffi_string(ucs: *mut u16, bytes: usize) -> OsString {
+    // Convert len in bytes to string length and do not count the null character
+    let len = bytes / crate::mem::size_of::<u16>() - 1;
+    let s = unsafe { crate::slice::from_raw_parts(ucs, len) };
+    OsString::from_wide(s)
+}
```

Comment:
This looks like an unfortunate API. You take a u16 array but size in bytes? Why not let the caller convert the length, so you don't have to decide on what to do with odd lengths?

Also, why does the length include the zero-character? What happens to intermediate zero-characters?

Lastly, `from_wide()` is lossy and converts to the replacement character, but given its use in file-system operations, this will eventually lead to files being inaccessible when you use the resulting string to access the file/directory again, since it no longer matches the name on the disk. I am unsure whether lossy conversions are really that useful for file-system operations.

Reply:
Well, the reason for using the size in bytes with null characters is because that's what most UEFI protocols return. Since this is an internal helper function, I moved the length calculation to this function.

We can also add a method for lossless conversion that may return an `Option<OsString>` or `Result<OsString>` if that would be better.

Reply:
The lossy nature of `from_wide()` should be fixed now since the windows implementation is lossless as far as I know.